File: src/components/PersonalPlan.js
// PersonalPlan.js
import React, { useState } from "react";
import "./PersonalPlan.css";

export default function PersonalPlan() {
  const [form, setForm] = useState({
    name: "",
    age: "",
    sex: "male",
    weight: "",
    height: "",
    goal: "weight loss",
    activity_level: "lightly active",
    experience: "beginner",
    diet_pref: "vegetarian",
    days_per_week: 4,
    preferred_time: "morning",
    health_conditions: ""
  });

  const [plan, setPlan] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");

  const handleChange = (e) => {
    const { name, value } = e.target;
    setForm(prev => ({ ...prev, [name]: value }));
  };

  const validate = () => {
    if (!form.name || !form.age || !form.weight || !form.height) {
      setError("Please fill name, age, weight and height.");
      return false;
    }
    if (form.age <= 12) {
      setError("This tool is meant for teen/adult users (13+).");
      return false;
    }
    setError("");
    return true;
  }

  const submit = async (e) => {
    e.preventDefault();
    if (!validate()) return;

    setLoading(true);
    setPlan(null);

    // Build payload; convert days_per_week to integer
    const payload = {
      ...form,
      days_per_week: parseInt(form.days_per_week, 10),
      health_conditions: form.health_conditions ? form.health_conditions.split(",").map(s => s.trim()) : []
    };

    try {
      const res = await fetch("http://127.0.0.1:5000/generate-plan", {
        method: "POST",
        headers: {"Content-Type": "application/json"},
        body: JSON.stringify(payload)
      });

      if (!res.ok) throw new Error("Network response was not ok");

      const data = await res.json();
      if (data.status !== "success") {
        setError("Failed to generate plan.");
      } else {
        setPlan(data.plan);
      }
    } catch (err) {
      console.error(err);
      setError("Could not connect to the backend. Is PersonalPlan.py running?");
    } finally {
      setLoading(false);
    }
  };

  return (
    <div className="personal-plan-container">
      <h2>Personal Plan Generator</h2>

      <form className="personal-plan-form" onSubmit={submit}>
        <div className="row">
          <input name="name" placeholder="Full name" value={form.name} onChange={handleChange} required />
          <input type="number" name="age" placeholder="Age" value={form.age} onChange={handleChange} required />
        </div>

        <div className="row">
          <select name="sex" value={form.sex} onChange={handleChange}>
            <option value="male">Male</option>
            <option value="female">Female</option>
            <option value="other">Other</option>
          </select>

          <input type="number" name="height" placeholder="Height (cm)" value={form.height} onChange={handleChange} required />
        </div>

        <div className="row">
          <input type="number" name="weight" placeholder="Weight (kg)" value={form.weight} onChange={handleChange} required />
          <select name="goal" value={form.goal} onChange={handleChange}>
            <option value="weight loss">Weight Loss</option>
            <option value="muscle gain">Muscle Gain</option>
            <option value="endurance">Endurance</option>
            <option value="flexibility">Flexibility</option>
            <option value="general fitness">General Fitness</option>
          </select>
        </div>

        <div className="row">
          <select name="activity_level" value={form.activity_level} onChange={handleChange}>
            <option value="sedentary">Sedentary</option>
            <option value="lightly active">Lightly active</option>
            <option value="moderately active">Moderately active</option>
            <option value="very active">Very active</option>
            <option value="extra active">Extra active</option>
          </select>

          <select name="experience" value={form.experience} onChange={handleChange}>
            <option value="beginner">Beginner</option>
            <option value="intermediate">Intermediate</option>
            <option value="advanced">Advanced</option>
          </select>
        </div>

        <div className="row">
          <select name="diet_pref" value={form.diet_pref} onChange={handleChange}>
            <option value="vegetarian">Vegetarian</option>
            <option value="non-vegetarian">Non-vegetarian</option>
          </select>

          <input type="number" name="days_per_week" min="1" max="7" value={form.days_per_week} onChange={handleChange}  placeholder="Days per week"/>
        </div>

        <div className="row">
          <select name="preferred_time" value={form.preferred_time} onChange={handleChange}>
            <option value="morning">Morning</option>
            <option value="afternoon">Afternoon</option>
            <option value="evening">Evening</option>
          </select>

          <input name="health_conditions" placeholder="Health conditions (comma separated)" value={form.health_conditions} onChange={handleChange} />
        </div>

        {error && <p className="error">{error}</p>}

        <button type="submit" className="primary-btn">Generate Plan</button>
      </form>

      {loading && <p className="loading">Generating plan — this may take a few seconds...</p>}

      {plan && (
        <div className="plan-result">
          <h3>{plan.name} — {plan.goal.toUpperCase()}</h3>
          <p className="small">
            Age: {plan.age} • Sex: {plan.sex} • Weight: {plan.weight_kg} kg • Height: {plan.height_cm} cm • BMI: {plan.BMI ?? "N/A"}
          </p>

          <div className="summary-grid">
            <div className="card">
              <h4>Calories</h4>
              <p>TDEE: {plan.TDEE_kcal ?? "N/A"} kcal/day</p>
              <p>Target: {plan.calorie_target_kcal} kcal/day</p>
            </div>

            <div className="card">
              <h4>Schedule</h4>
              <p>Days/week: {plan.days_per_week}</p>
              <p>Preferred time: {plan.preferred_time}</p>
            </div>

            <div className="card">
              <h4>Habits</h4>
              <ul>{plan.habits.map((h, i) => <li key={i}>{h}</li>)}</ul>
            </div>
          </div>

          <h4>Weekly Workouts</h4>
          <div className="week-grid">
            {plan.weekly_workouts.map((d, i) => (
              <div className="day-card" key={i}>
                <strong>{new Date(plan.weekly_meals[i].date).toLocaleDateString()}</strong>
                <p className="muted">{d.type === "workout" ? "Workout" : "Recovery"}</p>
                <p>{d.name}</p>
                <p className="muted">Duration: {d.duration_min} min • Intensity: {d.intensity}</p>
              </div>
            ))}
          </div>

          <h4>Daily Meals (sample)</h4>
          <div className="meals-grid">
            {plan.weekly_meals.map((day, idx) => (
              <div className="meal-card" key={idx}>
                <strong>{new Date(day.date).toLocaleDateString()}</strong>
                <ul>
                  {day.meals.map((m, i) => (
                    <li key={i}><span className="when">{m.when}:</span> {m.item} <span className="cal">({m.cal} kcal)</span></li>
                  ))}
                </ul>
              </div>
            ))}
          </div>

          <h4>Progress Guidelines</h4>
          <ul className="guidelines">
            {Object.entries(plan.weekly_progress_guidelines).map(([k, v]) => <li key={k}>{v}</li>)}
          </ul>

          <p className="small muted">{plan.notes}</p>
        </div>
      )}
    </div>
  );
}

File: src/components/PersonalPlan.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import mysql.connector
from werkzeug.security import generate_password_hash, check_password_hash
import random, json
from datetime import date, timedelta
from datetime import datetime, date

app = Flask(__name__)
CORS(app)

CORS(app, resources={r"/api/*": {"origins": "*"}})


# -----------------------------
# DATABASE CONNECTION
# -----------------------------
db = mysql.connector.connect(
    host="localhost",
    user="root",
    password="",
    database="fitwise",
    ssl_disabled=True
)
cursor = db.cursor(dictionary=True)

# =====================================================
# AUTH
# =====================================================
@app.route("/signup", methods=["POST"])
def signup():
    data = request.json or {}
    required = ["name", "email", "password", "age", "gender", "height_cm", "weight_kg"]
    if not all(data.get(k) for k in required):
        return jsonify({"error": "All fields required"}), 400

    cursor.execute("SELECT user_id FROM user WHERE email=%s", (data["email"],))
    if cursor.fetchone():
        return jsonify({"error": "Email exists"}), 409

    cursor.execute("""
        INSERT INTO user (name,email,password_hash,age,gender,height_cm,weight_kg, created_at)
        VALUES (%s,%s,%s,%s,%s,%s,%s, NOW())
    """, (
        data["name"],
        data["email"],
        generate_password_hash(data["password"]),
        data["age"],
        data["gender"],
        data["height_cm"],
        data["weight_kg"]
    ))
    db.commit()
    return jsonify({"message": "Signup successful"}), 201


@app.route("/login", methods=["POST"])
def login():
    data = request.json or {}
    cursor.execute("SELECT * FROM user WHERE email=%s", (data.get("email"),))
    user = cursor.fetchone()
    if not user or not check_password_hash(user["password_hash"], data.get("password", "")):
        return jsonify({"error": "Invalid credentials"}), 401

    cursor.execute("SELECT goal_type FROM goals WHERE user_id=%s ORDER BY created_at DESC LIMIT 1", (user["user_id"],))
    goal = cursor.fetchone()

    return jsonify({
        "user_id": user["user_id"],
        "name": user["name"],
        "age": user["age"],
        "gender": user["gender"],
        "height_cm": user["height_cm"],
        "weight_kg": user["weight_kg"],
        "goal": goal["goal_type"] if goal else None
    })

# =====================================================
# PROFILE + GOAL
# =====================================================
@app.route("/update-profile", methods=["POST"])
def update_profile():
    data = request.json or {}

    cursor.execute("UPDATE user SET height_cm=%s, weight_kg=%s WHERE user_id=%s",
                   (data["height_cm"], data["weight_kg"], data["user_id"]))
    cursor.execute("INSERT INTO goals (user_id, goal_type, created_at) VALUES (%s,%s,NOW())",
                   (data["user_id"], data["goal"]))
    db.commit()
    return jsonify({"message": "Profile updated"})


# =====================================================
# PLAN DATASETS
# =====================================================
WORKOUTS = {
    "weight loss": ["HIIT", "Brisk Walk", "Cycling"],
    "muscle gain": ["Push Pull Legs", "Upper Lower", "Full Body"],
    "general fitness": ["Mixed Cardio", "Mobility + Core"]
}

MEALS = {
    "vegetarian": ["Dal Rice", "Paneer Bowl", "Veg Upma"],
    "non-vegetarian": ["Chicken Rice", "Egg Omelette", "Fish Curry"]
}

HABITS = [
    "Drink 2-3L water",
    "Sleep 7-9 hours",
    "Stretch daily",
    "Track progress weekly"
]


@app.route("/generate-plan", methods=["POST"])
def generate_plan():
    data = request.json or {}
    required = ["name", "age", "sex", "height", "weight", "goal"]
    if not all(data.get(k) for k in required):
        return jsonify({"error": "Missing required fields"}), 400

    name = data["name"]
    age = int(data["age"])
    gender = data["sex"].lower()
    height = float(data["height"])
    weight = float(data["weight"])
    goal = data.get("goal", "general fitness").lower()
    diet_pref = data.get("diet_pref", "vegetarian")
    days = int(data.get("days_per_week", 3))
    preferred_time = data.get("preferred_time", "morning")

    bmi = round(weight / ((height / 100) ** 2), 1)
    bmr = (10 * weight) + (6.25 * height) - (5 * age) + (5 if gender == "male" else -161)
    tdee = int(bmr * 1.375)

    # Weekly workouts
    weekly_workouts = []
    for i in range(7):
        weekly_workouts.append({
            "day_index": i,
            "type": "workout" if i < days else "recovery",
            "name": random.choice(WORKOUTS.get(goal, WORKOUTS["general fitness"])),
            "duration_min": random.randint(30, 60),
            "intensity": "moderate"
        })

    # Weekly meals
    weekly_meals = []
    today = date.today()
    for i in range(7):
        weekly_meals.append({
            "date": (today + timedelta(days=i)).isoformat(),
            "meals": [
                {
                    "when": "Meal",
                    "item": random.choice(MEALS.get(diet_pref, MEALS["vegetarian"])),
                    "cal": 500
                }
            ]
        })

    plan = {
        "name": name,
        "age": age,
        "sex": gender,
        "weight_kg": weight,
        "height_cm": height,
        "BMI": bmi,
        "BMR_kcal": round(bmr),
        "TDEE_kcal": tdee,
        "calorie_target_kcal": tdee,
        "goal": goal,
        "days_per_week": days,
        "preferred_time": preferred_time,
        "weekly_workouts": weekly_workouts,
        "weekly_meals": weekly_meals,
        "habits": [
            "Drink 2–3L water daily",
            "Sleep 7–8 hours",
            "Warm up before workouts",
            "Stretch after exercise"
        ],
        "weekly_progress_guidelines": {
            "weight": "Track weekly",
            "strength": "Increase gradually",
            "recovery": "Rest when needed"
        },
        "notes": "This is an auto-generated plan. Adjust based on recovery."
    }

    cursor.execute("INSERT INTO personal_plans (name, goal, plan_json, created_at) VALUES (%s,%s,%s,NOW())",
                   (name, goal, json.dumps(plan)))
    db.commit()

    return jsonify({"status": "success", "plan": plan})

# ADMIN – FETCH USERS
@app.route("/admin/users", methods=["GET"])
def admin_users():
    cursor.execute("""
        SELECT user_id, name, email, age, gender, height_cm, weight_kg,
               IFNULL(created_at, NOW()) AS created_at
        FROM user
        ORDER BY user_id DESC
    """)
    rows = cursor.fetchall()
    # Ensure all required fields exist for frontend
    users = []
    for r in rows:
        users.append({
            "user_id": r["user_id"],
            "name": r["name"],
            "email": r["email"],
            "age": r["age"] or "",
            "gender": r["gender"] or "",
            "height_cm": r["height_cm"] or "",
            "weight_kg": r["weight_kg"] or "",
            "created_at": r["created_at"].isoformat() if isinstance(r["created_at"], (date,)) else str(r["created_at"]),
            "activity": {},
            "limitations": {"max_searches": 0}
        })
    return jsonify(users)


# ADMIN – FETCH PLANS
@app.route("/admin/plans", methods=["GET"])
def admin_plans():
    cursor.execute("""
        SELECT id, name, goal, created_at, plan_json
        FROM personal_plans ORDER BY id DESC
    """)
    rows = cursor.fetchall()
    plans = []
    for r in rows:
        plans.append({
            "id": r["id"],
            "name": r["name"],
            "goal": r["goal"],
            "created_at": r["created_at"].isoformat() if isinstance(r["created_at"], (date,)) else str(r["created_at"]),
            "plan": json.loads(r["plan_json"])
        })
    return jsonify(plans)

# =====================================================
# ADMIN – ADD & DELETE USERS/PLANS
# =====================================================
@app.route("/admin/add_user", methods=["POST"])
def admin_add_user():
    data = request.json or {}
    required = ["name", "email", "password", "age", "gender", "height_cm", "weight_kg", "limitations"]
    if not all(k in data for k in required):
        return jsonify({"error": "All fields required"}), 400

    cursor.execute("SELECT user_id FROM user WHERE email=%s", (data["email"],))
    if cursor.fetchone():
        return jsonify({"error": "Email exists"}), 409

    cursor.execute("""
        INSERT INTO user (name,email,password_hash,age,gender,height_cm,weight_kg, created_at)
        VALUES (%s,%s,%s,%s,%s,%s,%s,NOW())
    """, (
        data["name"], data["email"], generate_password_hash(data["password"]),
        data["age"], data["gender"], data["height_cm"], data["weight_kg"]
    ))
    db.commit()
    user_id = cursor.lastrowid

    return jsonify({
        "user_id": user_id,
        "name": data["name"],
        "email": data["email"],
        "age": data["age"],
        "gender": data["gender"],
        "height_cm": data["height_cm"],
        "weight_kg": data["weight_kg"],
        "activity": {},
        "limitations": data["limitations"]
    })


@app.route("/admin/user/<int:user_id>", methods=["DELETE"])
def delete_user(user_id):
    cursor.execute("DELETE FROM user WHERE user_id=%s", (user_id,))
    db.commit()
    return jsonify({"status": "deleted"})


@app.route("/admin/plan/<int:plan_id>", methods=["DELETE"])
def delete_plan(plan_id):
    cursor.execute("DELETE FROM personal_plans WHERE id=%s", (plan_id,))
    db.commit()
    return jsonify({"status": "deleted"})


if __name__ == "__main__":
    app.run(debug=True)



# ===========================
# DB MANAGER ROUTES
# ===========================

from datetime import datetime, date

# -----------------------------
# FETCH ALL USERS
# -----------------------------
@app.route("/admin/users", methods=["GET"])
def db_users():
    cursor.execute("""
        SELECT user_id, name, email, age, gender, height_cm, weight_kg,
               IFNULL(created_at, NOW()) AS created_at
        FROM user ORDER BY user_id DESC
    """)
    rows = cursor.fetchall()
    users = []
    for r in rows:
        users.append({
            "user_id": r["user_id"],
            "name": r["name"],
            "email": r["email"],
            "age": r["age"] or "",
            "gender": r["gender"] or "",
            "height_cm": r["height_cm"] or "",
            "weight_kg": r["weight_kg"] or "",
            "created_at": r["created_at"].isoformat() if isinstance(r["created_at"], (datetime, date)) else str(r["created_at"]),
            "activity": {},
            "limitations": {"max_searches": 0}
        })
    return jsonify(users)


# -----------------------------
# FETCH ALL PLANS
# -----------------------------
@app.route("/admin/plans", methods=["GET"])
def db_plans():
    cursor.execute("""
        SELECT id, name, goal, created_at, plan_json
        FROM personal_plans ORDER BY id DESC
    """)
    rows = cursor.fetchall()
    plans = []
    for r in rows:
        plans.append({
            "id": r["id"],
            "name": r["name"],
            "goal": r["goal"],
            "created_at": r["created_at"].isoformat() if isinstance(r["created_at"], (datetime, date)) else str(r["created_at"]),
            "plan": json.loads(r["plan_json"]) if r["plan_json"] else {}
        })
    return jsonify(plans)


# -----------------------------
# ADD USER
# -----------------------------
@app.route("/admin/add_user", methods=["POST"])
def db_add_user():
    data = request.json or {}
    required = ["name", "email", "password", "age", "gender", "height_cm", "weight_kg"]
    if not all(k in data and data[k] for k in required):
        return jsonify({"error": "All fields required"}), 400

    cursor.execute("SELECT user_id FROM user WHERE email=%s", (data["email"],))
    if cursor.fetchone():
        return jsonify({"error": "Email exists"}), 409

    cursor.execute("""
        INSERT INTO user (name,email,password_hash,age,gender,height_cm,weight_kg, created_at)
        VALUES (%s,%s,%s,%s,%s,%s,%s,NOW())
    """, (
        data["name"], data["email"], generate_password_hash(data["password"]),
        data["age"], data["gender"], data["height_cm"], data["weight_kg"]
    ))
    db.commit()
    user_id = cursor.lastrowid

    return jsonify({
        "user_id": user_id,
        "name": data["name"],
        "email": data["email"],
        "age": data["age"],
        "gender": data["gender"],
        "height_cm": data["height_cm"],
        "weight_kg": data["weight_kg"],
        "activity": {},
        "limitations": {"max_searches": 0}
    })


# -----------------------------
# DELETE USER
# -----------------------------
@app.route("/admin/user/<int:user_id>", methods=["DELETE"])
def db_delete_user(user_id):
    cursor.execute("DELETE FROM user WHERE user_id=%s", (user_id,))
    db.commit()
    return jsonify({"status": "deleted"})


# -----------------------------
# DELETE PLAN
# -----------------------------
@app.route("/admin/plan/<int:plan_id>", methods=["DELETE"])
def db_delete_plan(plan_id):
    cursor.execute("DELETE FROM personal_plans WHERE id=%s", (plan_id,))
    db.commit()
    return jsonify({"status": "deleted"})


